### DISCUSSION TITLE
A bufferline-like lualine component

### DISCUSSION THREAD

**Discussion by al-ce** [cbochs/grapple.nvim#167] [Show and tell]
Just wanted to share, I haven't used bufferline in a while but I liked the idea of seeing my tagged buffers at the top with a bit more information than the builtin provides. Though I wonder if I could achieve the same thing with existing Grapple API, or at least simplify this quite a bit,  but I couldn't figure it out in my initial attempts.

![screengrab-2024-05-30_13:01:32](https://github.com/cbochs/grapple.nvim/assets/23170004/9a1d4aec-c775-4eec-b0c4-3be1f91c7107)

That's the tabline with @mikesmithgh 's [git-prompt-string-lualine](https://github.com/mikesmithgh/git-prompt-string-lualine.nvim) followed by the grapple component, then the winbar underneath showing the current buffer.

```lua
M.grapple = {

  function()
    if not package.loaded["grapple"] then
      return "󰛣 "
    end
    local has_devicons, devicons = pcall(require, "nvim-web-devicons")
    local icon, hl = "", ""
    local tags = require("grapple").tags()
    if tags == nil or #tags == 0 then
      return "󰛣 "
    end
    local display = { "󰛢 " }
    for _, tag in ipairs(tags) do
      local path = tag.path
      if has_devicons then
        icon, hl = devicons.get_icon(path, string.match(path, "%a+$"), { default = true })
      end

      local filename = vim.fn.fnamemodify(path, ":t")
      local file_hl = path == vim.api.nvim_buf_get_name(0) and "%#Normal#" or "%#Comment#"
      table.insert(display, "%#" .. hl .. "#" .. icon)
      table.insert(display, file_hl .. filename)
    end
    table.insert(display, "%#Normal#" .. " 󰛢 ")
    return table.concat(display, " ")
  end,
  color = {
    fg = "#bbc2cf",
    bg = "#1a1b26",
  },
}
```

**Comment by AlexErrant** [comment]
[`akinsho/bufferline.nvim`](https://github.com/akinsho/bufferline.nvim) has a `name_formatter` option that I used:

```lua
        name_formatter = function(buf)
			local tags = require("grapple").tags()
            for k, v in ipairs(tags) do
                if v["path"] == buf["path"] then
                    return v["name"] .. " " .. buf["name"]
                end
            end
		end,
```

![image](https://github.com/user-attachments/assets/8040a0ea-ada3-4c99-87b8-ba9ebf44abc9)

I'm using named tags (Function Keys). My keymap uses `redrawtabline` to force a rerender when setting a tag, otherwise there's no immediate visual feedback.

```lua
{ "<F24>", "<cmd>Grapple tag name=F12<cr><cmd>redrawtabline<cr>", desc = "Set F12 tag" }, -- (Shift F12 is F24.)
```
